File: 2191/SortTheJumbledNumbers.java
/*
https://leetcode.com/problems/sort-the-jumbled-numbers/description/?envType=daily-question&envId=2024-07-24
*/

class CustomPair{
    public int mappedNumber;
    public int originalNumber;
    public int indexOfNumber;

    CustomPair(int mappedNumber, int originalNumber, int indexOfNumber){
        this.mappedNumber=mappedNumber;
        this.originalNumber=originalNumber;
        this.indexOfNumber=indexOfNumber;
    }
}

class CustomComparator implements Comparator<CustomPair>{

    @Override
    public int compare(CustomPair A, CustomPair B){
        if(A.mappedNumber==B.mappedNumber){
            return Integer.compare(A.indexOfNumber, B.indexOfNumber);
        }
        return Integer.compare(A.mappedNumber, B.mappedNumber);
    }
}


class Solution {

    public static String trimPrefixZeros(StringBuilder str){
        int lengthOfStr=str.length(), index=0;
        while(true){
            if(str.charAt(index)!='0') break;
            str.setCharAt(index, '#');
            index++;
            if(index>=lengthOfStr) break;
        }
        if(str.length()==1&&str.charAt(0)=='#') return "0";
        StringBuilder fnstr=new StringBuilder();
        for(int i=index; i<lengthOfStr; ++i){
            fnstr.append(str.charAt(i));
        }
        return String.valueOf(fnstr);
    }

    public int[] sortJumbled(int[] mapping, int[] nums) {
        ArrayList<CustomPair>listOfMappedNumbers=new ArrayList<>();
        int sz=nums.length;
        for(int i=0; i<sz; ++i){
            int currentNum=nums[i];
            String currentNumInString=String.valueOf(currentNum);
            StringBuilder mappedStringNumber=new StringBuilder();
            int lengthOfCurrentNumInString=currentNumInString.length();
            for(int j=0; j<lengthOfCurrentNumInString; ++j){
                mappedStringNumber.append(String.valueOf(mapping[currentNumInString.charAt(j)-'0']));
            }
            String trimmedString=trimPrefixZeros(mappedStringNumber);
            int newInt=Integer.parseInt(trimmedString);
            CustomPair newPair=new CustomPair(newInt, currentNum, i);
            listOfMappedNumbers.add(newPair);
        }
        Collections.sort(listOfMappedNumbers, new CustomComparator());
        int[] resultantList=new int[listOfMappedNumbers.size()];
        for(int i=0; i<listOfMappedNumbers.size(); ++i){
            resultantList[i]=listOfMappedNumbers.get(i).originalNumber;
        }
        return resultantList;
    }
}

File: 2/2000s/21s/2191/SortTheJumbledNumbers.cpp
/*
https://leetcode.com/problems/sort-the-jumbled-numbers/description/?envType=daily-question&envId=2024-07-24
*/

class CustomPair {
public:
    int var1;
    int var2;
    int var3;

public:
    CustomPair(int var1, int var2, int var3){
        this->var1=var1;
        this->var2=var2;
        this->var3=var3;
    }
};

class CustomComparator{
    public:

    bool operator()(const CustomPair A, const CustomPair B){
        if(A.var1==B.var1){
            return A.var3<B.var3;
        }
        return A.var1<B.var1;
    }
};

class Solution {
public:

    void trimPrexficZeros(string &str){
        int sz=str.size(), idx=0;
        while(true){
            if(str[idx]!='0') break;
            str[idx]='#';
            idx++;
            if(idx>=sz) break;
        }
        str.erase(remove(str.begin(), str.end(), '#'), str.end());
        if(str.size()==0) str="0";
    }

    vector<int> sortJumbled(vector<int>& mapping, vector<int>& nums) {
        vector<CustomPair>vec;
        int sz=nums.size();
        for(int i=0; i<sz; ++i){
            int curnum=nums[i];
            string tochk=to_string(curnum);
            string tofrm="";
            for(auto &ch: tochk){
                tofrm+=to_string(mapping[(ch-'0')]);
            }
            trimPrexficZeros(tofrm);
            int nwint=stoi(tofrm);
            CustomPair pr(stoi(tofrm), curnum, i);
            vec.push_back(pr);
        }
        sort(vec.begin(), vec.end(), CustomComparator());
        vector<int>res;
        for(auto &vals: vec){
            res.push_back(vals.var2);
        }
        return res;
    }
};